packages/api/lambdas/sf-event-sqs-to-db-records/write-granules.js

```diff
@@ -14,43 +19,196 @@ const {
   getMessageGranules,
   getGranuleStatus,
 } = require('@cumulus/message/Granules');
+const {
+  getMessageProvider,
+} = require('@cumulus/message/Providers');
+const {
+  getMessageWorkflowStartTime,
+  getWorkflowDuration,
+} = require('@cumulus/message/workflows');
 
+const FileUtils = require('../../lib/FileUtils');
+const {
+  getExecutionProcessingTimeInfo,
+  getGranuleTimeToArchive,
+  getGranuleTimeToPreprocess,
+  getGranuleProductVolume,
+} = require('../../lib/granules');
+const {
+  parseException,
+} = require('../../lib/utils');
 const Granule = require('../../models/granules');
 
-const writeGranuleViaTransaction = async ({
+const generateGranuleRecord = async ({
   cumulusMessage,
   granule,
   collectionCumulusId,
   providerCumulusId,
   executionCumulusId,
   pdrCumulusId,
+  files,
+  processingTimeInfo = {},
+  cmrUtils = CmrUtils,
+  timestamp = Date.now(),
+  updatedAt = Date.now(),
+}) => {
+  const {
+    granuleId,
+    cmrLink,
+    published = false,
+  } = granule;
+
+  const workflowStartTime = getMessageWorkflowStartTime(cumulusMessage);
+  const temporalInfo = await cmrUtils.getGranuleTemporalInfo(granule);
+
+  return {
+    granule_id: granuleId,
+    status: getGranuleStatus(cumulusMessage, granule),
+    cmr_link: cmrLink,
+    error: parseException(cumulusMessage.exception),
+    published,
+    created_at: new Date(workflowStartTime),
+    updated_at: new Date(updatedAt),
+    timestamp: new Date(timestamp),
+    // Duration is also used as timeToXfer for the EMS report
+    duration: getWorkflowDuration(workflowStartTime, timestamp),
+    product_volume: getGranuleProductVolume(files),
+    time_to_process: getGranuleTimeToPreprocess(granule),
+    time_to_archive: getGranuleTimeToArchive(granule),
+    collection_cumulus_id: collectionCumulusId,
+    provider_cumulus_id: providerCumulusId,
+    execution_cumulus_id: executionCumulusId,
+    pdr_cumulus_id: pdrCumulusId,
+    // Temporal info from CMR
+    beginning_date_time: temporalInfo.beginningDateTime,
+    ending_date_time: temporalInfo.endingDateTime,
+    production_date_time: temporalInfo.productionDateTime,
+    last_update_date_time: temporalInfo.lastUpdateDateTime,
+    // Processing info from execution
+    processing_start_date_time: processingTimeInfo.processingStartDateTime,
+    processing_end_date_time: processingTimeInfo.processingEndDateTime,
+  };
+};
+
+const generateFileRecord = ({ file, granuleCumulusId }) => ({
+  ...translateApiFiletoPostgresFile(file),
+  granule_cumulus_id: granuleCumulusId,
+});
+
+const generateFileRecords = async ({
+  files,
+  granuleCumulusId,
+}) => files.map((file) => generateFileRecord({ file, granuleCumulusId }));
+
+const writeFilesViaTransaction = async ({
```

Comment:
I'm not super opposed, though then this would be inconsistent with the code for writing executions and PDRs. I'm kinda fine with `upsert` being an implementation detail of how the `write` is done